File: bronze/greedy/race.java
// INPUT
// k meters race, num of diff X's
// diff X's: no more than X speed at the end

import java.io.*;
import java.util.*;

public class Main {

    // Solves one query: given distance K and ending speed limit X
    public static int solve(int dist, int minSpeed) {
        int leftTravel = 0;   // distance while speeding up
        int rightTravel = 0;  // distance while slowing down
        int timeUsed = 0;

        // Try increasing speed forever until we exceed distance
        for (int currSpeed = 1; ; currSpeed++) {

            // Speeding up phase
            leftTravel += currSpeed;
            timeUsed++;

            if (leftTravel + rightTravel >= dist) {
                return timeUsed;
            }

            // Slowing down phase (only allowed if currSpeed >= X)
            if (currSpeed >= minSpeed) {
                rightTravel += currSpeed;
                timeUsed++;

                if (leftTravel + rightTravel >= dist) {
                    return timeUsed;
                }
            }
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("race.in"));
        PrintWriter out = new PrintWriter(new FileWriter("race.out"));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int K = Integer.parseInt(st.nextToken()); // distance
        int N = Integer.parseInt(st.nextToken()); // number of queries

        for (int i = 0; i < N; i++) {
            int X = Integer.parseInt(br.readLine());
            out.println(solve(K, X));
        }

        out.close();
        br.close();
    }
}

File: bronze/greedy/race.py

def min_time(K, X):
    """
    Returns the minimum number of seconds needed
    to run at least K meters, starting at speed 1,
    and ending with speed >= X.
    """

    curr_distance = 0
    slow_down_distance = 0

    time = 0
    speed = 1
    
    # Phase 1: speed < X (only speed up)
    while speed < X:
        curr_distance += speed
        time += 1
        if curr_distance >= K:
            return time
        speed += 1

    # Phase 2: speed >= X (speed up + slow down mirror)
    while True:
        curr_distance += speed
        time += 1
        if curr_distance + slow_down_distance >= K:
            return time

        slow_down_distance += speed
        time += 1
        if curr_distance + slow_down_distance >= K:
            return time

        speed += 1
    '''
    while True:
        # consider the new distance
        curr_distance += current_speed
        time += 1

        # see if total possible distance is enough
        if curr_distance + slow_down_distance >= K:
            return time

        # if above speed limit, then consider distance if slow down
        if current_speed >= X:
            slow_down_distance += current_speed
            time += 1

            if curr_distance + slow_down_distance >= K:
                return time

        # increase speed
        current_speed += 1
    '''

with open("race.in", "r") as fin:
    K, N = map(int, fin.readline().split())
    queries = [int(fin.readline()) for _ in range(N)]

with open("race.out", "w") as fout:
    for X in queries:
        fout.write(str(min_time(K, X)) + "\n")